bitbots_localization/src/localization.cpp

```diff
@@ -67,92 +53,124 @@ rcl_interfaces::msg::SetParametersResult Localization::onSetParameters(
   robot_state_distribution_start_left_.reset(new RobotStateDistributionStartLeft(
-      random_number_generator_, std::make_pair(config_->field_x, config_->field_y)));
-
+      random_number_generator_, std::make_pair(config_.field.size.x, config_.field.size.y)));
   robot_state_distribution_start_right_.reset(new RobotStateDistributionStartRight(
-      random_number_generator_, std::make_pair(config_->field_x, config_->field_y)));
-
-  robot_state_distribution_left_half_.reset(
-      new RobotStateDistributionLeftHalf(random_number_generator_, std::make_pair(config_->field_x, config_->field_y)));
-
+      random_number_generator_, std::make_pair(config_.field.size.x, config_.field.size.y)));
+  robot_state_distribution_left_half_.reset(new RobotStateDistributionLeftHalf(
+      random_number_generator_, std::make_pair(config_.field.size.x, config_.field.size.y)));
   robot_state_distribution_right_half_.reset(new RobotStateDistributionRightHalf(
-      random_number_generator_, std::make_pair(config_->field_x, config_->field_y)));
-
+      random_number_generator_, std::make_pair(config_.field.size.x, config_.field.size.y)));
   robot_state_distribution_position_.reset(
-      new RobotStateDistributionPosition(random_number_generator_, config_->initial_robot_x, config_->initial_robot_y));
-
+      new RobotStateDistributionPosition(random_number_generator_, config_.field.initialization.single_pose.x,
+                                         config_.field.initialization.single_pose.y));
   robot_state_distribution_pose_.reset(new RobotStateDistributionPose(
-      random_number_generator_, config_->initial_robot_x, config_->initial_robot_y, config_->initial_robot_t));
-
-  resampling_.reset(new pf::ImportanceResampling<RobotState>(true, config_->min_resampling_weight));
-
-  RCLCPP_INFO(this->get_logger(), "Trying to initialize particle filter...");
-  reset_filter(config_->init_mode);
-
-  if (first_configuration_) {
-    first_configuration_ = false;
-    reset_service_ = this->create_service<bl::srv::ResetFilter>(
-        "reset_localization", std::bind(&Localization::reset_filter_callback, this, _1, _2));
-    pause_service_ = this->create_service<bl::srv::SetPaused>(
-        "pause_localization", std::bind(&Localization::set_paused_callback, this, _1, _2));
+      random_number_generator_, config_.field.initialization.single_pose.x, config_.field.initialization.single_pose.y,
+      config_.field.initialization.single_pose.t));
+
+  // Create the resampling strategy
+  resampling_.reset(
+      new pf::ImportanceResampling<RobotState>(true, config_.particle_filter.weighting.particle_reset_weight));
+
+  // Check if we need to create a new particle filter or if we can update the existing one (keeping the particle states)
+  if (!robot_pf_) {
+    // Create new particle filter
+    robot_pf_.reset(new particle_filter::ParticleFilter<RobotState>(
```

Comment:
where was this PF set before your changes?

Reply:
In the reset filter function. Because we create a new filter every time we reset the localization. And this is done also at startup.